File: ch7OOD/Ques7_12.cpp
#include <list>
#include <vector>
#include <string>
#include <iostream>
using namespace std;

class MyHashTable {
private:
    vector<list<string>> ht;
    unsigned sz = 0;   

public:
    MyHashTable(int size) {
        sz = size;
        ht = vector<list<string>>(size);
    }

    unsigned hash(string s) {
        int charSum = 0;
        for (auto c : s)
            charSum += c;
        cout << "charSum: " << charSum << endl;
        return (s.length() + charSum) % sz;
    }

    unsigned size() {
        return sz;
    }

    bool containsKey(string key) {
        if (ht[hash(key)].size() > 0)
            return true;
    
        return false;
    }

    void print() {
        int i = 0;
        for (auto l : ht) {
            cout << i << ")";
            for(auto s : l) {
                cout<<(s + " -> ");
            }
            cout<<endl;
            ++i;
        }
    }

    void put(string key) {
        ht[hash(key)].push_back(key);
    }
};

int main(void) {
    MyHashTable mht(10);
    string sa[] = {"a","ab","abc", "abcd", "abcde", "abcdef"};
    for (string s : sa) {
        mht.put(s);
    }
    mht.print();
}

File: ch07OOD/Ques7_12.java
import java.util.*;

class MyHashTable
{
	private ArrayList<LinkedList<String>> ht;
	private int sz = 0;

	public MyHashTable(int size){
		sz = size;
        ht = new ArrayList<LinkedList<String>>(size);
        for (int i = 0; i < size; ++i)
            ht.add(new LinkedList<String>());
	}

	public final int hash(String s)
	{
		int charSum = 0;
		for (char c : s.toCharArray()) {
			charSum += c;
		}
		return Math.abs(s.hashCode() + s.length() + charSum) % sz;
	}

	public final int size() {
		return sz;
	}

	public final boolean containsKey(String key) {
		if (ht.get(hash(key)).size() > 0) {
			return true;
		}
		return false;
	}

	public final void print() {
		int i = 0;
		for (LinkedList<String> l : ht) {
			System.out.print(i + ") ");
			for (var s : l) {
				System.out.print((s + " -> "));
			}
			System.out.print("\n");
			++i;
		}
	}

	public final void put(String key) {
		ht.get(hash(key)).addLast(key);
	}
}

public class Ques7_12 {
	public static void main(String[] args) {
		MyHashTable mht = new MyHashTable(10);
		String[] sa = {"a", "ab", "abc", "abcd", "abcde", "abcdef"};
		for (String s : sa){
			mht.put(s);
		}
		mht.print();
	}
}
